### DISCUSSION TITLE
AD equivalent to ParsedMaterial?

### DISCUSSION THREAD

**Discussion by LukeFisherUKAEA** [idaholab/moose#19132] [Q&A Modules: General] [answered]
Hello,

Is there an existing Automatic Differentiation equivalent to ParsedMaterial within the framework?  I am moving to an ADKernel-based simulation but need to solve the issue of ADKernels requiring ADMaterialProperties.  Alternatively, this may require modification of existing Material blocks to declare ADMaterialProperty values?

Thanks for the help.

**Comment by ABallisat** [comment, marked as answer]
ADParsedMaterial does exist but it is not explicitly in the docs, it is defined in here:
https://github.com/idaholab/moose/blob/master/framework/src/materials/ParsedMaterial.C

I have found this in a few places throughout MOOSE, in general if it does not appear in the docs I check the source code as it may already be there.

**Reply by GiudGiud** [reply]
@lindsayad do you think we should add some standardized text to class that is doubled up between AD and non AD?
Or prefer doubling up the documentation?
Is there an issue where people discussed this? i looked at the AD PR and the AD issue https://github.com/idaholab/moose/issues/5658, could not find it

**Reply by lindsayad** [reply]
The `bool is_ad` idiom developed separately from all the primary AD work. It's been an idiom that's really taken off because of developers who want to maintain both regular and AD versions of classes with minimal effort, which I applaud them for. But I'm not surprised that there is no mention of this in the original AD issue or PR because I never envisioned it really.

I think it would be good to somehow indicate in the documentation all the concrete types that arise from a class template. That could definitely help our users

**Reply by lindsayad** [reply]
@LukeFisherUKAEA `ADParsedMaterial` should be the answer here as @ABallisat suggested. Please let us know if there are any issues with that.

**Reply by lindsayad** [reply]
In case it's of interest, `ADParsedMaterial` was added in https://github.com/idaholab/moose/pull/15331 in reference to https://github.com/idaholab/moose/issues/15207

**Reply by GiudGiud** [reply]
ok raising an issue to keep track of this
https://github.com/idaholab/moose/issues/19137

**Reply by LukeFisherUKAEA** [reply]
Thanks all - this is what I was looking for.
